whylog/config/rule.py

```diff
@@ -57,6 +65,265 @@ def get_causes_parsers(self):
+        return {
+            group_type: {
+                InvestigationStep.LEFT_BOUND: primary_group_value,
+                InvestigationStep.RIGHT_BOUND: primary_group_value
+            }
+        }
+
+    def _aggregate_constraints(self):
+        """
+        This method groups delta constraints by equal base parser
+        """
+        parser_with_constraints = defaultdict(list)
+        for constraint in self._constraints:
+            if constraint['name'] in self.DELTA_CONSTRAINTS:
+                base_parser = constraint['clues_groups'][1][0]
+                parser_with_constraints[base_parser].append(constraint)
+        return parser_with_constraints
+
+    def _is_primary_key_constraint(self, clues_groups):
+        """
+        This method checks that groups connected by delta constraint belongs to primary key
+        """
+        base_parser_number = clues_groups[1][0]
+        depended_parser_number = clues_groups[0][0]
+        base_parser_group_number = clues_groups[1][1]
+        depended_group_number = clues_groups[0][1]
+        return self._is_group_primary(base_parser_group_number, base_parser_number) and \
+               self._is_group_primary(depended_group_number, depended_parser_number)
+
+    def _is_group_primary(self, parser_group_number, parser_number):
+        if parser_number == self.EFFECT_NUMBER:
+            return self._effect.is_primary_key(parser_group_number)
+        return self._causes[parser_number - 1].is_primary_key(parser_group_number)
+
+    def _calculate_parser_bounds(self, base_parser_number, params, group_type, parser_ranges):
+        """
+        This method calculate depended parser search range basing od base parser search range.
+        """
+        max_delta = params.get('max_delta')
+        min_delta = params.get('min_delta')
+        left_bound, right_bound = self._get_base_bounds(
+            base_parser_number, group_type, parser_ranges
+        )
+        converter = CONVERTION_MAPPING[group_type]
+        new_left_bound = converter.switch_by_delta(left_bound, max_delta, "max")
+        new_right_bound = converter.switch_by_delta(right_bound, min_delta, "min")
+        return {
+            group_type: {
+                InvestigationStep.LEFT_BOUND: new_left_bound,
+                InvestigationStep.RIGHT_BOUND: new_right_bound
+            }
+        }
+
+    @classmethod
+    def _get_base_bounds(cls, base_parser_number, group_type, parser_ranges):
+        base_parser_bounds = parser_ranges[base_parser_number][group_type]
+        left_bound = base_parser_bounds[InvestigationStep.LEFT_BOUND]
+        right_bound = base_parser_bounds[InvestigationStep.RIGHT_BOUND]
+        return left_bound, right_bound
+
+    def create_ranges_for_unconnected_parsers(
+        self, effect_clues, effect_primary_group, effect_group_type, parser_ranges
+    ):
+        """
+        Create maximal search range for parsers which are unreachable by delta constraints.
```

Comment:
What does it mean that parser is unreachable by constraint? I don't understand. Please improve this docstring.